### PR TITLE
Refactor the context classes

### PR DESCRIPTION
**Pull request by ste93cry**
In this PR I refactored the `*Context` classes by optimizing them to store data into typed properties rather than plain arrays: this allowed me to also remove their dependency on the [OptionsResolver](https://symfony.com/doc/current/components/options_resolver.html) component of Symfony and to drop the base `Context` class because of a few reasons:

- The `extra` field is not really a context as it's a top-level field of the event payload. Also, since the field can basically contain anything it didn't really make sense to store its data in anything else than a plain array in the `Event` class
- The `user` field is not really a context for the same reason as above, therefore having a `UserContext` class was semantically wrong

I also extracted the intialization of the default values of the contexts into an `EnvironmentIntegration` integration so that it's easier to solve issues like #985 by just disabling the integration.

### CODE REVIEW

File: UPGRADE-3.0.md [hunk 1]
```diff
@@ -37,3 +37,59 @@
...
+- The `OsContext` and `RuntimeContext` classes do not implement anymore the `ArrayAccess`, `IteratorAggregate` and `JsonSerializable` interfaces
+- The following methods have been removed from the `OsContext` and `RuntimeContext` classes:
+  - `*Context::merge()`
+  - `*Context::setData()`
+  - `*Context::replaceData()`
+  - `*Context::clear()`
+  - `*Context::isEmpty()`
+  - `*Context::toArray()`
+- Removed the `UserContext` class, use `UserDataBag` instead
+- The signature of the `RuntimeContext` class changed to require a bunch of parameters in place of an array of initial data
```
Comment (Jean85) [inline UPGRADE-3.0.md:51]: It also became `final`

File: UPGRADE-3.0.md [hunk 2]
```diff
@@ -37,3 +37,59 @@
 - The signature of the `HttpClientFactory::__construct()` method changed to accept instances of the PSR-17 factories in place of Httplug's ones
 - The signature of the `DefaultTransportFactory::__construct()` method changed to accept instances of the PSR-17 factories in place of Httplug's ones
 - The signature of the `GzipEncoderPlugin::__construct()` method changed to accept an instance of the `Psr\Http\Message\StreamFactoryInterface` interface only
+- Removed the `TagsContext`, `ExtraContext` and `Context` classes, data is now stored in a plain array
```
Comment (Jean85) [inline UPGRADE-3.0.md:40]: `TagContext` no longer does merge, always overrides right?

**Review by Jean85 — CHANGES_REQUESTED** [review]

File: UPGRADE-3.0.md (hunk 2 above)
Comment (ste93cry) [reply UPGRADE-3.0.md:40]: `TagsContext` used to provide a `merge()` function, but now both the `Event` and `Scope` classes provide a setter that imho is quite self-explanatory. Of course when applying a scope on an event the tags still get merged

**Review by Jean85 — APPROVED** [review]

**Review by stayallive — APPROVED** [review]
🚢

**Pull request merged**